datafusion/expr/src/expr.rs

```diff
@@ -983,10 +983,290 @@ impl PartialOrd for Expr {
 }
 
 impl Expr {
-    /// Returns the name of this expression as it should appear in a schema. This name
-    /// will not include any CAST expressions.
-    pub fn display_name(&self) -> Result<String> {
-        create_name(self)
+    /// Returns the name for schema / field that is different from Display
+    /// Most of the expressions are the same as Display
+    /// Those are the main difference
+    /// 1. Alias, where excludes expression
+    /// 2. Cast / TryCast, where takes expression only
```

Comment:
Here is a suggesiton for how to improve this documentation:

```suggestion
    /// The name of the column (field) that this `Expr` will produce.
    ///
    /// For example, for a projection (e.g. `SELECT <expr>`) the resulting arrow
    /// [`Schema`] will have a field with this name.
    ///
    /// Note that the resulting string is subtlety different than the `Display`
    /// representation for certain `Expr`. Some differences:
    ///
    /// 1. [`Expr::Alias`], which shows only the alias itself
    /// 2. [`Expr::Cast`] / [`Expr::TryCast`], which only displays the expression
    ///
    /// # Example
    /// ```
    /// # use datafusion_expr::{col, lit};
    /// let expr = col("foo").eq(lit(42));
    /// assert_eq!("foo = Int32(42)", expr.schema_name().unwrap());
    ///
    /// let expr = col("foo").alias("bar").eq(lit(11));
    /// assert_eq!("bar = Int32(11)", expr.schema_name().unwrap());
    /// ```
    ///
    /// [`Schema`]: arrow::datatypes::Schema
```